### PR TITLE
Use `LogitsProcessor` for `transformers` integration

### LINKED ISSUE #806: Update the `transformers` integration
In the vein of #782 and #772 we should refactor the `transformers` integration to use logits processor.

We will keep the custom sampling loop, but via a `Outlines` model that wraps `transformers` models. We should be able to remove `torch` and `transformers` as default dependencies as a result.

### PR DESCRIPTION
**Pull request by lapp0**
Fixes #806

Related: https://github.com/outlines-dev/outlines/issues/789

Requesting comment on whether we should unify the logits processor in this PR as described in the last section of this message. @brandonwillard @rlouf 

### Problem

`transformers` has two issues in its `generate` integration:
- if `input_ids` is empty, it fails
- it has an inconsistent implementation with other models, it is the only model which doesn't use logits processors

### Solution

Implement logits processors for transformers.

#### Proposed Additional Work

We have four implementations of logits processors which do the same thing. Their implementations have drifted and now there are growing differences in their implementation details. We should simplify this and use a single logits processor for all models, ensuring a bug fix or enhancement for one logits processor applies to all models.

- Additionally, this minimizes the changes necessary for https://github.com/outlines-dev/outlines/issues/918  minimal
- Furthermore, this issue only would need to be fixed in one module with a singular implementation: https://github.com/outlines-dev/outlines/issues/855

This WIP PR shows generalized `logits_processors.py` which aims to be compatible with all inference engines, along with a modified `generate/regex.py` incorporating the new singular logits processor implementation.

### CODE REVIEW

**Comment by lapp0** [comment]
closed in favor of https://github.com/outlines-dev/outlines/pull/966
